server/src/main/java/org/opensearch/action/admin/cluster/repositories/put/PutRepositoryRequest.java

```diff
@@ -67,12 +70,22 @@ public class PutRepositoryRequest extends AcknowledgedRequest<PutRepositoryReque
 
     private Settings settings = EMPTY_SETTINGS;
 
+    private Boolean encrypted;
+    private CryptoSettings cryptoSettings;
+
     public PutRepositoryRequest(StreamInput in) throws IOException {
         super(in);
         name = in.readString();
         type = in.readString();
         settings = readSettingsFromStream(in);
         verify = in.readBoolean();
+
+        if (in.getVersion().onOrAfter(Version.V_3_0_0)) {
+            encrypted = in.readOptionalBoolean();
```

Comment:
Why is this an optional boolean? An optional boolean has three states, so what does `null` mean?

Reply:
It is marked as optional to ensure backward compatibility with existing integrations. We don't want existing repository registrations to start failing from this version. We can treat a null encrypted field as a non encrypted repo. If we move it to `readBoolean` then it will become a breaking change. To add more info, stream based constructors are typically used in node to node communication.

Reply:
Optional doesn't give you backwards compatibility; when invoking `writeOptionalBoolean` you're still unconditionally writing a byte to the stream and the reader must unconditionally read this byte as well. The `onOrAfter` version check gives you backwards compatibility. I think you can just write a regular boolean and default the field to false if not explicitly set when constructing an instance. I think you can also use a primitive boolean type in the class and avoid the `null` case all together.

Reply:
I wasn't referring to version backwards compatibility but the ability for an existing api to work without any change with the new field introduction during let's say upgrading the cluster. Let me check this further if I can move it to non optional field without breaking anything.

Reply:
I looked into this and following are the two cases :
1. We keep encrypted field `Boolean` and use `writeBoolean` or `readBoolean` methods instead of their optional counterparts. In this case lets say you have 2 nodes cluster. You trigger a put repository request on non active cluster manager node (say node2) without encrypted field. Node2 will try to form the request for cluster manager node using `writeBoolean` and it will fail with NPE as it expects a non-null boolean value. If we had used an optional Boolean then it would have done a null check first, written a byte value of 2 and sent request to cluster manager node. On cluster manager node, again it would require optional method to read since it needs to handle the possibility of null value. Basically, to keep a non-primitive field and to use non-optional methods, users will have to change their contracts.
2. We keep primitive version this field. If you do this then you don't have a null case anymore which means that in all the get calls users will start seeing encrypted field. Again, if there are strict contracts on user side to read these responses, then they will start failing as well because of the unknown field. By keeping non-primitive fields users can selectively change contracts for encryption and other users continue to stay unchanged. 

Correct me if I am missing something here which you are trying to point!

Reply:
Got it, you're right that in order to preserve the "not set" state in the repository settings then we'll need the optional boolean to communicate that third possible state.